net/tstun/wrap.go

```diff
@@ -615,63 +670,102 @@ func (c *natV4Config) selectSrcIP(oldSrc, dst netip.Addr) netip.Addr {
 	return oldSrc
 }
 
-// natV4ConfigFromWGConfig generates a natV4Config from nm.
-// If v4 NAT is not required, it returns nil.
-func natV4ConfigFromWGConfig(wcfg *wgcfg.Config) *natV4Config {
+// natConfigFromWGConfig generates a natConfig from nm.
+// If NAT is not required, it returns nil.
+func natConfigFromWGConfig(wcfg *wgcfg.Config) *natConfig {
 	if wcfg == nil {
 		return nil
 	}
-	nativeAddr := findV4(wcfg.Addresses)
-	if !nativeAddr.IsValid() {
+
+	nativeAddr4 := findV4(wcfg.Addresses)
+	nativeAddr6 := findV6(wcfg.Addresses)
+	if !nativeAddr4.IsValid() && !nativeAddr6.IsValid() {
 		return nil
 	}
+
 	var (
-		rt           table.RoutingTableBuilder
-		dstMasqAddrs map[key.NodePublic]netip.Addr
-		listenAddrs  set.Set[netip.Addr]
+		rt4, rt6                     table.RoutingTableBuilder
+		dstMasqAddrs4, dstMasqAddrs6 map[key.NodePublic]netip.Addr
+		listenAddrs4, listenAddrs6   set.Set[netip.Addr]
 	)
 
 	// When using an exit node that requires masquerading, we need to
 	// fill out the routing table with all peers not just the ones that
 	// require masquerading.
-	exitNodeRequiresMasq := false // true if using an exit node and it requires masquerading
+	var exitNodeRequiresMasq4, exitNodeRequiresMasq6 bool // true if using an exit node and it requires masquerading
 	for _, p := range wcfg.Peers {
 		isExitNode := slices.Contains(p.AllowedIPs, tsaddr.AllIPv4()) || slices.Contains(p.AllowedIPs, tsaddr.AllIPv6())
-		if isExitNode && p.V4MasqAddr != nil && p.V4MasqAddr.IsValid() {
-			exitNodeRequiresMasq = true
-			break
+		if isExitNode {
+			if p.V4MasqAddr != nil && p.V4MasqAddr.IsValid() {
+				exitNodeRequiresMasq4 = true
+			}
+			if p.V6MasqAddr != nil && p.V6MasqAddr.IsValid() {
+				exitNodeRequiresMasq6 = true
+			}
+			if exitNodeRequiresMasq4 && exitNodeRequiresMasq6 {
+				break
+			}
 		}
 	}
+
 	for i := range wcfg.Peers {
 		p := &wcfg.Peers[i]
-		var addrToUse netip.Addr
+		var addrToUse4 netip.Addr
 		if p.V4MasqAddr != nil && p.V4MasqAddr.IsValid() {
-			addrToUse = *p.V4MasqAddr
-			mak.Set(&listenAddrs, addrToUse, struct{}{})
-		} else if exitNodeRequiresMasq {
-			addrToUse = nativeAddr
+			addrToUse4 = *p.V4MasqAddr
+			mak.Set(&listenAddrs4, addrToUse4, struct{}{})
+		} else if exitNodeRequiresMasq4 {
+			addrToUse4 = nativeAddr4
 		} else {
 			continue
```

Comment:
note: this fast-path case with the `continue` is unproven in tests. the tests only cover that the natcfg lookup result is correct, they do not prove that the fast-path is taken. This branch can be removed and the tests still pass.